cpp/include/cudf/io/orc_metadata.hpp

```diff
@@ -25,31 +25,236 @@
+};
+
+/**
+ * @brief Statistics for floating point columns.
+ */
+struct double_statistics : minmax_statistics<double>, sum_statistics<double> {
+  static constexpr statistics_type type = statistics_type::DOUBLE;
+};
+
+/**
+ * @brief Statistics for string columns.
+ *
+ * The `minimum` and `maximum` are the first and last elements, respectively, in lexicographical
+ * order. The `sum` is the total length of elements in the column.
+ * Note: According to ORC specs, the sum should be signed, but pyarrow uses unsigned value
+ */
+struct string_statistics : minmax_statistics<std::string>, sum_statistics<uint64_t> {
+  static constexpr statistics_type type = statistics_type::STRING;
+};
+
+/**
+ * @brief Statistics for boolean columns.
+ *
+ * The `count` array includes the count of `false` and `true` values.
+ */
+struct bucket_statistics {
+  static constexpr statistics_type type = statistics_type::BUCKET;
+  std::vector<uint64_t> _count;
+
+  auto count(size_t index) const { return &_count.at(index); }
+};
+
+/**
+ * @brief Statistics for decimal columns.
+ */
+struct decimal_statistics : minmax_statistics<std::string>, sum_statistics<std::string> {
+  static constexpr statistics_type type = statistics_type::DECIMAL;
+};
+
+/**
+ * @brief Statistics for date(time) columns.
+ */
+struct date_statistics : minmax_statistics<int32_t> {
+  static constexpr statistics_type type = statistics_type::DATE;
+};
+
+/**
+ * @brief Statistics for binary columns.
+ *
+ * The `sum` is the total number of bytes across all elements.
+ */
+struct binary_statistics : sum_statistics<int64_t> {
+  static constexpr statistics_type type = statistics_type::BINARY;
+};
+
+/**
+ * @brief Statistics for timestamp columns.
+ *
+ * The `minimum` and `maximum` min/max elements in the column, as the number of milliseconds since
+ * the UNIX epoch. The `minimum_utc` and `maximum_utc` are the same values adjusted to UTC.
+ */
+struct timestamp_statistics : minmax_statistics<int64_t> {
+  static constexpr statistics_type type = statistics_type::TIMESTAMP;
+  std::unique_ptr<int64_t> _minimum_utc;
+  std::unique_ptr<int64_t> _maximum_utc;
+
+  auto has_minimum_utc() const { return _minimum_utc != nullptr; }
+  auto has_maximum_utc() const { return _maximum_utc != nullptr; }
+  auto minimum_utc() const { return _minimum_utc.get(); }
+  auto maximum_utc() const { return _maximum_utc.get(); }
+};
+
+namespace orc {
+// forward declare the type that ProtobufReader uses. The `cudf::io::column_statistics` objects,
+// returned from `read_parsed_orc_statistics`, are constructed from
+// `cudf::io::orc::column_statistics` objects that `ProtobufReader` initializes.
+struct column_statistics;
+}  // namespace orc
+
+/**
+ * @brief Contains per-column ORC statistics.
+ *
+ * All columns can have the `number_of_values` statistics. Depending on the data type, a column can
+ * have additional statistics, accessible through `type_specific_stats` accessor.
+ */
+class column_statistics {
```

Comment:
@harrism you left a comment about the `column_statistics` design that got resolved as the code moved (I think).
The `column_statistics` class with many `unique_ptr` members is now internal. The API returns this `column_statistics`, which has the viod* + type enum implementation. I think the class has a decent API (see `type_specific_stats` below), given what it holds. Please take a look if you missed it previously.